File: native/macos/Sources/ACCore/Wait.swift
import Foundation
import ApplicationServices
import AppKit

// MARK: - Wait & Poll

class Wait {

    /// Wait for a fixed duration
    static func waitMs(ms: Int) -> [String: Any] {
        Thread.sleep(forTimeInterval: Double(ms) / 1000.0)
        return ["ok": true, "waited_ms": ms]
    }

    /// Wait for an element to exist/be visible, or disappear
    static func waitForElement(
        ref: String,
        hidden: Bool,
        enabled: Bool,
        timeout: Int,
        refMap: [String: AXUIElement]
    ) -> (result: [String: Any]?, error: RPCResponse?) {

        let deadline = Date().addingTimeInterval(Double(timeout) / 1000.0)

        while Date() < deadline {
            if let element = refMap[ref] {
                if hidden {
                    // Wait for element to disappear — it still exists, keep waiting
                } else if enabled {
                    // Wait for element to become enabled
                    var enabledRef: CFTypeRef?
                    AXUIElementCopyAttributeValue(element, kAXEnabledAttribute as CFString, &enabledRef)
                    if (enabledRef as? Bool) == true {
                        return (["ok": true, "ref": ref, "state": "enabled"], nil)
                    }
                } else {
                    // Element exists — success
                    return (["ok": true, "ref": ref, "state": "visible"], nil)
                }
            } else {
                if hidden {
                    // Element gone — success for --hidden mode
                    return (["ok": true, "ref": ref, "state": "hidden"], nil)
                }
            }
            Thread.sleep(forTimeInterval: 0.2) // Poll every 200ms
        }

        return (nil, RPCResponse.error(id: 0, code: RPCErrorCode.timeout,
            message: "Timeout waiting for element: \(ref)"))
    }

    /// Wait for an app to launch
    static func waitForApp(
        name: String,
        timeout: Int
    ) -> (result: [String: Any]?, error: RPCResponse?) {
        let deadline = Date().addingTimeInterval(Double(timeout) / 1000.0)

        while Date() < deadline {
            for app in NSWorkspace.shared.runningApplications {
                if app.localizedName?.lowercased() == name.lowercased() && app.isFinishedLaunching {
                    return (["ok": true, "app": name, "pid": Int(app.processIdentifier)], nil)
                }
            }
            Thread.sleep(forTimeInterval: 0.2)
        }

        return (nil, RPCResponse.error(id: 0, code: RPCErrorCode.timeout,
            message: "Timeout waiting for app: \(name)"))
    }

    /// Wait for a window with a given title
    static func waitForWindow(
        title: String,
        timeout: Int,
        windowManager: WindowManager
    ) -> (result: [String: Any]?, error: RPCResponse?) {
        let deadline = Date().addingTimeInterval(Double(timeout) / 1000.0)

        while Date() < deadline {
            let windows = windowManager.listWindows()
            for win in windows {
                if let winTitle = win["title"] as? String,
                   winTitle.lowercased().contains(title.lowercased()) {
                    return (["ok": true, "window": win], nil)
                }
            }
            Thread.sleep(forTimeInterval: 0.2)
        }

        return (nil, RPCResponse.error(id: 0, code: RPCErrorCode.timeout,
            message: "Timeout waiting for window: \(title)"))
    }

    /// Wait for text to appear (or disappear) in the active window
    static func waitForText(
        text: String,
        gone: Bool,
        timeout: Int,
        grabbedWindow: String?,
        windowManager: WindowManager,
        snapshotBuilder: SnapshotBuilder
    ) -> (result: [String: Any]?, error: RPCResponse?) {
        let deadline = Date().addingTimeInterval(Double(timeout) / 1000.0)

        while Date() < deadline {
            let (snapResult, _) = snapshotBuilder.build(
                windowRef: grabbedWindow,
                windowManager: windowManager
            )

            if let snapResult = snapResult,
               let elements = snapResult["elements"] as? [[String: Any]] {
                let found = searchForText(elements, text: text)

                if gone && !found {
                    return (["ok": true, "text": text, "state": "gone"], nil)
                } else if !gone && found {
                    return (["ok": true, "text": text, "state": "found"], nil)
                }
            }

            Thread.sleep(forTimeInterval: 0.3)
        }

        return (nil, RPCResponse.error(id: 0, code: RPCErrorCode.timeout,
            message: "Timeout waiting for text: \(text)"))
    }

    private static func searchForText(_ elements: [[String: Any]], text: String) -> Bool {
        let lowerText = text.lowercased()
        for el in elements {
            if let label = el["label"] as? String, label.lowercased().contains(lowerText) {
                return true
            }
            if let value = el["value"] as? String, value.lowercased().contains(lowerText) {
                return true
            }
            if let children = el["children"] as? [[String: Any]] {
                if searchForText(children, text: lowerText) { return true }
            }
        }
        return false
    }
}

File: native/windows/ACCore/Wait.cs
using System.Diagnostics;
using System.Windows.Automation;

namespace ACCore;

public class WaitHelper
{
    private readonly Dictionary<string, AutomationElement> _refMap;
    private readonly WindowManager _windowManager;

    public WaitHelper(Dictionary<string, AutomationElement> refMap, WindowManager windowManager)
    {
        _refMap = refMap;
        _windowManager = windowManager;
    }

    public object WaitMs(int ms)
    {
        Thread.Sleep(ms);
        return new { ok = true };
    }

    public object WaitForElement(string refStr, bool hidden = false, bool enabled = false, int timeout = 10000)
    {
        var start = Stopwatch.StartNew();

        while (start.ElapsedMilliseconds < timeout)
        {
            if (_refMap.TryGetValue(refStr, out var element))
            {
                try
                {
                    if (hidden)
                    {
                        // Wait for element to disappear
                        if (element.Current.IsOffscreen || element.Current.BoundingRectangle.IsEmpty)
                            return new { ok = true };
                    }
                    else if (enabled)
                    {
                        if (element.Current.IsEnabled)
                            return new { ok = true };
                    }
                    else
                    {
                        // Element exists and is visible
                        if (!element.Current.IsOffscreen)
                            return new { ok = true };
                    }
                }
                catch
                {
                    // Element went stale — if waiting for hidden, that counts
                    if (hidden) return new { ok = true };
                }
            }
            else if (hidden)
            {
                // Ref not in map — element is gone
                return new { ok = true };
            }

            Thread.Sleep(200);
        }

        throw new ACException(ErrorCodes.Timeout,
            $"Timed out waiting for element {refStr} (hidden={hidden}, enabled={enabled})");
    }

    public object WaitForApp(string name, int timeout = 10000)
    {
        var start = Stopwatch.StartNew();

        while (start.ElapsedMilliseconds < timeout)
        {
            var procs = Process.GetProcessesByName(name);
            foreach (var proc in procs)
            {
                try
                {
                    if (proc.MainWindowHandle != IntPtr.Zero)
                        return new { ok = true, process_id = proc.Id };
                }
                catch { }
            }
            Thread.Sleep(200);
        }

        throw new ACException(ErrorCodes.Timeout, $"Timed out waiting for app: {name}");
    }

    public object WaitForWindow(string title, int timeout = 10000)
    {
        var start = Stopwatch.StartNew();

        while (start.ElapsedMilliseconds < timeout)
        {
            var windows = _windowManager.ListWindows();
            var match = windows.FirstOrDefault(w =>
                w.Title.Contains(title, StringComparison.OrdinalIgnoreCase));
            if (match != null)
                return new { ok = true, window = match.Ref };

            Thread.Sleep(200);
        }

        throw new ACException(ErrorCodes.Timeout, $"Timed out waiting for window: {title}");
    }

    public object WaitForText(string text, bool gone, SnapshotBuilder snapshotBuilder,
        AutomationElement? windowElement, WindowInfo? windowInfo, int timeout = 10000)
    {
        var start = Stopwatch.StartNew();

        while (start.ElapsedMilliseconds < timeout)
        {
            if (windowElement != null && windowInfo != null)
            {
                var snapshot = snapshotBuilder.Build(windowElement, windowInfo);
                var found = ContainsText(snapshot.Elements, text);

                if (gone ? !found : found)
                    return new { ok = true };
            }

            Thread.Sleep(300);
        }

        throw new ACException(ErrorCodes.Timeout,
            $"Text \"{text}\" {(gone ? "did not disappear" : "not found")} within {timeout}ms");
    }

    private bool ContainsText(List<ElementInfo> elements, string text)
    {
        foreach (var el in elements)
        {
            if (el.Label?.Contains(text, StringComparison.OrdinalIgnoreCase) == true)
                return true;
            if (el.Value?.Contains(text, StringComparison.OrdinalIgnoreCase) == true)
                return true;
            if (el.Children != null && ContainsText(el.Children, text))
                return true;
        }
        return false;
    }
}
